server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java

```diff
@@ -123,7 +124,7 @@ private void handlePrimaryResult(final PrimaryResultT primaryResult) {
             // is valid for this replication group. If we would sample in the reverse, the global checkpoint might be based on a subset
             // of the sampled replication group, and advanced further than what the given replication group would allow it to.
             // This would entail that some shards could learn about a global checkpoint that would be higher than its local checkpoint.
-            final long globalCheckpoint = primary.globalCheckpoint();
+            final long globalCheckpoint = primary.computedGlobalCheckpoint();
```

Comment:
I'm a bit in trouble with this one. `computedGlobalCheckpoint()` returns the last known global checkpoint on primary which is potentially not yet persisted. Still, we communicate this value to the replicas which are going to take it as the source of truth. Isn't that a problem? I tend to think that `primary.globalCheckpoint()` is more appropriate here but maybe I'm missing something.

Reply:
The global checkpoint is just derived information from the replication group (i.e. the minimum of the local checkpoint of all in-sync shard copies), it does not really matter whether that information is persisted on the primary first. The same applies to other replicated information. Indexing operations for example are not fsynced on the primary either before they are replicated to replicas (see `AsyncAfterWriteAction`). The main thing is that the primary establishes what should be done, and then everyone in the replication group (including the primary) follows those instructions. In case anyone fails to follow those instructions (incl. the primary itself), they're failed.

Reply:
It makes sense, thanks for the detailed explanation.